solutions/usaco-694.mdx

```diff
@@ -104,5 +104,84 @@ int main() {
 ```
 
 </CPPSection>
+	
+<JavaSection>
+```java
+import java.io.*;
+import java.util.*;
+
+public class HoofPaperScissors {
+	static int[] moves;
+
+	public static void main (String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new FileReader("hps.in"));
+		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("hps.out")));
+
+		StringTokenizer inputLine = new StringTokenizer(br.readLine());
+		int numGames = Integer.parseInt(inputLine.nextToken());
+		int maximumSwitches = Integer.parseInt(inputLine.nextToken());
+		// dp[game][switches][type] = maximum number of wins
+		// at that game with at most that amount
+		// of switches and that with that type active
+		moves = new int[numGames + 1];
+		for (int x = 1; x <= numGames; x++) {
+			char c = br.readLine().charAt(0);
+			// set H as 0, P as 1, S as 2
+			if (c == 'H') {
+				moves[x] = 0;
+			}
+			else if (c == 'P') {
+				moves[x] = 1;
+			}
+			else {
+				moves[x] = 2;
+			}
+		}
+
+		int[][][] dp = new int[numGames + 1][maximumSwitches + 2][3];  
+		int maximum = 0;
+
+		for (int currentGame = 1; currentGame <= numGames; currentGame++) {  // for each game
+			// for each switch amount
+			for (int numberOfSwitches = 1; numberOfSwitches < maximumSwitches + 2; numberOfSwitches++) {  
+				// try switching to the three possibilities
+				int add1 = 0;
+				if (moves[currentGame] == 0) {
+					add1 = 1;
+				}
+				// take the maximum of the game before + add (win or lose)
+				// and switching from the other two options 
+				dp[currentGame][numberOfSwitches][0] = Math.max(dp[currentGame - 1][numberOfSwitches][0] 
+				+ add1, Math.max(dp[currentGame - 1][numberOfSwitches - 1][1] + add1, 
+				dp[currentGame - 1][numberOfSwitches - 1][2] + add1));
```

Comment:
this line splitting looks kind of weird, but these lines are kind of long so i'm not sure what the best way to format it is? anyone else have ideas?

Reply:
yeah it basically takes the maximum of the three matrix values in the dp so the line is that long. i could make the variable names shorter but then they won't be as descriptive and it will still be over the limit

Reply:
maybe i could get an exemption for this one because it is a special case taking the max of three values in a dp?

Reply:
long lines are fine, but the way there are split onto multiple lines looks kind of weird?? not sure the best way to split them though...

Reply:
if this was c++ you could probably just macro each dp position